src/tools/getdns_server_mon.c

```diff
@@ -0,0 +1,1211 @@
+#include <stddef.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
+#include <openssl/bio.h>
+
+#include <getdns/getdns.h>
+#include <getdns/getdns_extra.h>
+
+#define APP_NAME "getdns_server_mon"
+
+#define RTT_CRITICAL_MS                 250
+#define RTT_WARNING_MS                  500
+
+#define CERT_EXPIRY_CRITICAL_DAYS       7
+#define CERT_EXPIRY_WARNING_DAYS        14
+
+#define DEFAULT_LOOKUP_NAME             "getdnsapi.net"
+#define DEFAULT_LOOKUP_TYPE             GETDNS_RRTYPE_AAAA
+
+#define EDNS0_PADDING_CODE              12
+
+#define EXAMPLE_PIN "pin-sha256=\"E9CZ9INDbd+2eRQozYqqbQ2yXLVKB9+xcprMF+44U1g=\""
+
+/* Plugin exit values */
+typedef enum {
+        EXIT_OK = 0,
+        EXIT_WARNING,
+        EXIT_CRITICAL,
+        EXIT_UNKNOWN,
+        EXIT_USAGE              /* Special case - internal only. */
+} exit_value_t;
+
+/* Plugin verbosity values */
+typedef enum {
+        VERBOSITY_MINIMAL = 0,
+        VERBOSITY_ADDITIONAL,
+        VERBOSITY_CONFIG,
+        VERBOSITY_DEBUG
+} verbosity_t;
+
+static struct test_info_s
+{
+        getdns_context *context;
+
+        /* Output control */
+        bool monitoring;
+        FILE *errout;
+        verbosity_t verbosity;
+
+        /* Test config info */
+        bool fail_on_dns_errors;
+} test_info;
+
+static int get_rrtype(const char *t)
+{
+        char buf[1024] = "GETDNS_RRTYPE_";
+        uint32_t rrtype;
+        long int l;
+        size_t i;
+        char *endptr;
+
+        if (strlen(t) > sizeof(buf) - 15)
+                return -1;
+        for (i = 14; *t && i < sizeof(buf) - 1; i++, t++)
+                buf[i] = *t == '-' ? '_' : toupper(*t);
+        buf[i] = '\0';
+
+        if (!getdns_str2int(buf, &rrtype))
+                return (int)rrtype;
+
+        if (strncasecmp(buf + 14, "TYPE", 4) == 0) {
+                l = strtol(buf + 18, &endptr, 10);
+                if (!*endptr && l >= 0 && l < 65536)
+                        return l;
+        }
+        return -1;
+}
+
+static const char *rcode_text(int rcode)
+{
+        const char *text[] = {
+                "OK",
+                "FORMERR",
+                "SERVFAIL",
+                "NXDOMAIN",
+                "NOTIMP",
+                "REFUSED"
+        };
+
+        if ((size_t) rcode >= sizeof(text) / sizeof(text[0]))
+                return "(?)";
+        else
+                return text[rcode];
+}
+
+/* Thanks to:
+ * https://zakird.com/2013/10/13/certificate-parsing-with-openssl
+ */
+static bool extract_cert_expiry(const unsigned char *data, size_t len, time_t *t)
+{
+        X509 *cert = d2i_X509(NULL, &data, len);
+        if (!cert)
+                return false;
+
+        ASN1_TIME *not_after = X509_get_notAfter(cert);
+
+        /*
+         * Use ASN1_TIME_diff to get a time delta between now and expiry.
+         * This is much easier than trying to parse the time.
```

Comment:
Absolutely, however ASN1_TIME_diff is only available with OpenSSL >= 1.0.2, hence travis tests fail because they use ubuntu 14.04 which has openssl 1.01f.

What to do now?  Not build tool for OpenSSL < 1.02 or not build too, without expiry testing for OpenSSL < 1.02 or copy some function that converts ASN1_TIME to time_t.